feat: Add global shortcuts

tuxinal opened the pull request:
~~i'm using electron's globalShortcuts api here which doesn't have a way to detect tell the difference between keyup and keydown. this means the push to talk/mute/deafen/etc actions don't really work. we'd probably have to switch to something else at some point~~
currently implementing [venbind](https://github.com/tuxinal/venbind) to use instead of electron's api 

resolves #18

tuxinal committed:
Add global shortcuts

sunnniee commented on src/renderer/patches/keybinds.ts:
```diff
@@ -0,0 +1,45 @@
+        {
+            find: "[kb store] KeybindStore",
+            replacement: {
+                match: /(inputEventRegister\(parseInt\((.{1,2})\),(.{1,2}),(.{1,2}),(.{1,2})\);else\{)([^;]*;[^;]*;.{1,2}\.keyup&&.{1,2}\.bindGlobal\(\(0,(.{1,2}\.toString)\))/,
```
you should use `\i` instead of `.{1,2}` to match variables. eslint will yell at you as it wasn't patched here unlike in main vencord; just tell it to shot up with an `// eslint-ignore`

sunnniee commented on src/renderer/patches/keybinds.ts:
```diff
@@ -0,0 +1,45 @@
+        },
+        {
+            find: "[kb store] KeybindStore",
+            replacement: {
```
the replacement variable can be an array to not have to find the same module twice

tuxinal committed:
Minimize regex
convert id to number
enable desktop only shortcut actions

tuxinal commented:
Okay so electron's globalShortcuts api is not at all a good fit for this. the keyup/keydown events are rather crucial for discord's functionality. and also it "consumes" inputs (i have no idea what the correct term for this is. basically if you set anything as a keybind nothing else can use it, for example if you have the tab key as push to talk tab won't do anything everywhere else).

Vendicated commented:
yes. as mentioned in the issue you linked and as discussed on discord, using a native library would be better. however all the current options don't seem to support wayland at all, so we will most likely have to write our own module

tuxinal committed:
cli based toggles
Co-Authored-By: exhq <infidelLOler@gmail.com>

tuxinal committed:
Add keyup/keydown support for the cli
TEMPORARY: show keybind ID in the keybind page

klhrt commented:
Could you do a quick write-up of how this works? E.g. is `vesktop --keybind 2` supposed to execute the keybind of id 2 set inside the app, or is the cli used to attach a bind to a key? It's pretty difficult for me to parse the code to figure out how to use it.

tuxinal commented:
when you make a new keybind in discord asigns it an id. currently i have the id for each keybind written next to the keybind in the client. 
you can use this id in the cli as: `(vesktop) --keybind (id)`
or if it's ptt you can add a `keyup`/`keydown` thing at the end of it.

alterNERDtive commented:
I was just about to write a feature request about something like this … so :+1:

D3SOX commented on src/renderer/patches/keybinds.ts:
```diff
@@ -0,0 +1,45 @@
+        {
+            find: "[kb store] KeybindStore",
+            replacement: {
+                // WHY IS THE RADIX SPEICIFIED
```
Just FYI I know why the radix is specified now

From https://github.com/denysdovhan/wtfjs#parseint-is-a-bad-guy:

> 💡 Explanation: If the input string begins with "0", radix is eight (octal) or 10 (decimal). Exactly which radix is chosen is implementation-dependent. ECMAScript 5 specifies that 10 (decimal) is used, but not all browsers support this yet. For this reason always specify a radix when using parseInt.

tuxinal committed:
update patch

klhrt commented:
this doesn't seem to work on my setup on ubuntu with i3+x11, unless I still don't get how to use it. not sure if it's just not compatible with my setup or i'm just stupid, but if it doesn't work due to a window manager or display server problem it's something to look into.

if i set a toggle mute keybind with an arbitrary key and it gets assigned an id of 3, running `vesktop --keybind 3` doesn't toggle mute.

tuxinal committed:
fix shortcut string function

tuxinal commented:
@klhrt try this commit

klhrt commented:
```
UnhandledPromiseRejectionWarning: Error: Script failed to execute, this normally means an error was thrown. Check the renderer console for the error.
    at node:electron/js2c/renderer_init:2:16470
    at IpcRendererInternal.<anonymous> (node:electron/js2c/renderer_init:2:10723)
    at IpcRendererInternal.emit (node:events:519:28)
    at Object.onMessage (node:electron/js2c/renderer_init:2:8837)
(node:1605447) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 33)
```
Whenever I run the keybind command, no response in-app and this error pops up. Not sure if the error is responsible/related though.

for reference this is the resulting printout from the terminal where i run the keybind command:
```
~> vesktop --keybind 2
Checking for beta autoupdate feature for deb/rpm distributions
Found package-type: deb
Checking for update
~>
```

klhrt commented:
> @klhrt ATM the command is `vesktop -- --keybind id` 🥴

and it works! thanks

tuxinal committed:
add venbind. currently only linux x64 on x11

tuxinal commented:
it would be nice if someone could test this commit on their system. the binary is technically built on nix and idk if that affects anything

tuxinal committed:
separate venbind copy function

tuxinal committed:
fix unregistering not working

rbutera commented:
anyone know how hard it would be to get this to work on Hyprland?

tuxinal commented:
you'd have to implement xdg-desktop-portal's GlobalShortucts api for hyprland on their portal implementation (which they might have implemented already i'm not sure)
i'm working on the vesktop side of things as of right now and it's proving somewhat difficult. no one really has a complete implementation of it

condmaker commented:
Doesn't appear to be working on Hyprland or wayland in general. I can't even start the instance. Here's the full stack trace if that helps:

```
thread '<unnamed>' panicked at src/lib.rs:16:44:
called `Result::unwrap()` on an `Err` value: Message("todo")
stack backtrace:
   0:     0x7e1f3eae23d5 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h3692694645b1bb6a
   1:     0x7e1f3eb0661b - core::fmt::write::h5131d80b4c69b88d
   2:     0x7e1f3eadff8f - std::io::Write::write_fmt::h1fb327a7d8b0eb36
   3:     0x7e1f3eae21ae - std::sys_common::backtrace::print::h998d75b840f75a73
   4:     0x7e1f3eae36f9 - std::panicking::default_hook::{{closure}}::h18ec7fe6a38b9da0
   5:     0x7e1f3eae349a - std::panicking::default_hook::hfb3f22c2e4075a6a
   6:     0x7e1f3eae3b93 - std::panicking::rust_panic_with_hook::h51af00bcb4660c4e
   7:     0x7e1f3eae3a74 - std::panicking::begin_panic_handler::{{closure}}::h39f76aa863fbe8ce
   8:     0x7e1f3eae2899 - std::sys_common::backtrace::__rust_end_short_backtrace::h4d10fc2251b89840
   9:     0x7e1f3eae37a7 - rust_begin_unwind
  10:     0x7e1f3e9cea13 - core::panicking::panic_fmt::h319840fcbcd912ef
  11:     0x7e1f3e9ceea6 - core::result::unwrap_failed::haccb9aaa604e1e21
  12:     0x7e1f3e9dfa8c - std::sys_common::backtrace::__rust_begin_short_backtrace::hfc2e22cd1f52e7a4
  13:     0x7e1f3e9dd1ff - core::ops::function::FnOnce::call_once{{vtable.shim}}::h3001236f03d016af
  14:     0x7e1f3eae5ecb - std::sys::pal::unix::thread::Thread::new::thread_start::h3b8e81128811868f
  15:     0x7e1fd0c4439d - <unknown>
  16:     0x7e1fd0cc949c - <unknown>
  17:                0x0 - <unknown>
[arRPC > ipc] listening at /run/user/1000/discord-ipc-0
[arRPC > websocket] listening on 6463
[arRPC > process] started
thread '<unnamed>' panicked at src/lib.rs:20:44:
called `Result::unwrap()` on an `Err` value: Message("todo")
stack backtrace:
   0:     0x7e1f3eae23d5 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h3692694645b1bb6a
   1:     0x7e1f3eb0661b - core::fmt::write::h5131d80b4c69b88d
   2:     0x7e1f3eadff8f - std::io::Write::write_fmt::h1fb327a7d8b0eb36
   3:     0x7e1f3eae21ae - std::sys_common::backtrace::print::h998d75b840f75a73
   4:     0x7e1f3eae36f9 - std::panicking::default_hook::{{closure}}::h18ec7fe6a38b9da0
   5:     0x7e1f3eae349a - std::panicking::default_hook::hfb3f22c2e4075a6a
   6:     0x7e1f3eae3b93 - std::panicking::rust_panic_with_hook::h51af00bcb4660c4e
   7:     0x7e1f3eae3a74 - std::panicking::begin_panic_handler::{{closure}}::h39f76aa863fbe8ce
   8:     0x7e1f3eae2899 - std::sys_common::backtrace::__rust_end_short_backtrace::h4d10fc2251b89840
   9:     0x7e1f3eae37a7 - rust_begin_unwind
  10:     0x7e1f3e9cea13 - core::panicking::panic_fmt::h319840fcbcd912ef
  11:     0x7e1f3e9ceea6 - core::result::unwrap_failed::haccb9aaa604e1e21
  12:     0x7e1f3e9d82bc - venbind::js::napi_register_keybind::hee504851f7cfee42
  13:     0x5c87ef3a456c - <unknown>
fatal runtime error: failed to initiate panic, error 5
[1]    79116 IOT instruction (core dumped)  RUST_BACKTRACE=full ./vesktop
```

ImLvna commented:
> Doesn't appear to be working on Hyprland or wayland in general. I can't even start the instance. Here's the full stack trace if that helps:
looks like venbind isnt setup for wayland yet, id just wait
https://github.com/tuxinal/venbind/blob/885e0dd7658bb2e25f1ad45f8d6581b0e330febb/src/linux.rs#L30

tuxinal commented:
oh none of the wayland support stuff is pushed yet. currently venbind only supports linux x11

OrHy3 commented:
Over the weekend I started writing C library (based on `libdbus`) to ease interactions with XDG GlobalShortcuts. The said library is [libgsbus](https://github.com/OrHy3/libgsbus), it's in a very basic release state, but it's complete.
As of now it lacks documentation and a build system, but it's a single source-header files pair, the latter being commented.

If anyone is willing to test it around, it'll be greatly appreciated.

checkraisefold committed:
Changes for windows support

checkraisefold commented:
As an update for anyone with the PR watched: further work on Venbind has begun to support Windows and Wayland (through xdg-desktop-portal DBus), and is underway.

checkraisefold commented:
An alternative to XDP if at all possible may need to be explored at some point for Venbind considering there is currently 3 different (2 confirmed) bugs blocking global shortcuts in KDE and Hyprland, which afaik are the only two actually usable global shortcuts XDP implementations

https://github.com/hyprwm/xdg-desktop-portal-hyprland/pull/310
https://bugs.kde.org/show_bug.cgi?id=492992

OrHy3 commented:
Last time I worked with GlobalShortcuts on KDE it reused the previous session just fine (also tested right now).
Regarding your pull request, seems like it already got merged.

I honestly think the D-Bus portal is the way to go, mostly because will likely be the next standard for Wayland, but also because (from what I get) there is no intention here to use libevdev or similar input-permission related methods, even if it's just gonna be temporary.

checkraisefold commented:
> Last time I worked with GlobalShortcuts on KDE it reused the previous session just fine (also tested right now). Regarding your pull request, seems like it already got merged.
> 
> I honestly think the D-Bus portal is the way to go, mostly because will likely be the next standard for Wayland, but also because (from what I get) there is no intention here to use libevdev or similar input-permission related methods, even if it's just gonna be temporary.

Yeah, the hyprland xdg PR got merged in 10 minutes apparently. However, the KDE bug was blocking testing of Venbind's XDG implementation by the Venbind maintainer - I couldn't test it on Hyprland because of that bug, but I'll give it a try with the latest commit of xdg-desktop-portal-hyprland.

StayBlue commented:
I think it's important to note that the Venbind maintainer hasn't had any Github activity over the past couple of months, so it might be necessary to make a hard fork of it if he doesn't reply soon. That decision is probably best left in the hands of @checkraisefold, as he is the one who is bringing back momentum into this PR.

checkraisefold commented:
> I think it's important to note that the Venbind maintainer hasn't had any Github activity over the past couple of months, so it might be necessary to make a hard fork of it if he doesn't reply soon. That decision is probably best left in the hands of @checkraisefold, as he is the one who is bringing back momentum into this PR.

On that note, the maintainer is active in the Vencord Discord server #vesktop-development channel (where I have been very active in regards to looking at the Wayland/XDG issues, I pinged you in there). Shouldn't need to make a hard fork, I think they'll be able to merge the XDG pr soon once I test it on Hyprland.

checkraisefold commented:
Basic Windows and Wayland (via xdg-desktop-portal) support have been merged into Venbind! 🥳 

There won't be another tagged version of Venbind until it is at least somewhat usable (confirmed with maintainer) - key release events aren't implemented and there's issues with the existing implementation that need to be fixed. Once that's done, and Darwin support is done (shouldn't be too hard - libuiohook has it inbuilt) then Venbind should be pretty ready, and then everything's on the JavaScript side to implement.

tuxinal committed:
Add all available actions to globalshortcuts on startup

tuxinal committed:
Show error on platforms using XDP

tuxinal committed:
Add keyup keydown support
update venbind

tuxinal committed:
Merge branch 'main' into main

StayBlue commented:
Hello! I know this is a draft PR, and it's still a work in progress, but I tried this on Windows by setting a keybind in the settings and doing:
```ps
.\Vesktop.exe -- --keybind 2
```
I also tried doing:
 ```ps
.\Vesktop.exe --keybind 2
```
And nothing happened when I tried to press the assigned keybind. Also, for whatever reason, trying to set a bind from your mouse in the settings doesn't appear to register.

tuxinal committed:
Update lockfile properly
Github merge ui is 💀

tuxinal committed:
update findbycode query

tuxinal commented:
turns out an update caused issues with registering new keybinds. can you try the latest commit?

Snowy-Fluffy commented:
Is it possible to do something similar on Linux?

Snowy-Fluffy commented:
> Is it possible to do something similar on Linux?

Ah, I noticed.

tuxinal committed:
Fix Push To Talk
you love half line fixes

RinLovesYou commented:
working beautifully here on KDE (wayland)

StayBlue commented:
It works with the latest commit. The only thing I'd want on my wish list is a way to have mouse bindings, even if I have to do it manually since Discord's in-client keybinding manager doesn't pick it up because of Electron.

tuxinal commented:
can you not manually bind your mouse through the cli with the latest commit?

baanan commented:
Hello, is glibc 2.39 the minimum required for venbind? I'm running Pop! OS 22.04 with a glib version of 2.35, and when I try to run the branch, it explains that venbind needs glibc 2.39 to run (full error below). I also tried building venbind myself, and the error persisted.
<details>
  <summary>Full Error</summary>

```

> vesktop@1.5.5 start /home/thate/Documents/utilities/Vesktop
> pnpm build && electron .


> vesktop@1.5.5 build /home/thate/Documents/utilities/Vesktop
> tsx scripts/build/build.mts

APPIMAGE env is not defined, current application is not an AppImage
Vesktop v1.5.5
checkForUpdatesAndNotify called, downloadPromise is null
(node:52635) UnhandledPromiseRejectionWarning: Error: /lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.39' not found (required by /home/thate/Documents/utilities/Vesktop/static/dist/venbind-linux-x86_64.node)
    at process.func [as dlopen] (node:electron/js2c/node_init:2:2559)
    at Module._extensions..node (node:internal/modules/cjs/loader:1602:18)
    at Object.func [as .node] (node:electron/js2c/node_init:2:2559)
    at Module.load (node:internal/modules/cjs/loader:1295:32)
    at Module._load (node:internal/modules/cjs/loader:1111:12)
    at c._load (node:electron/js2c/node_init:2:16955)
    at Module.require (node:internal/modules/cjs/loader:1318:19)
    at require (node:internal/modules/helpers:179:18)
    at Vr (VCDMain:87:48639)
    at g0 (VCDMain:87:48718)
(Use \`electron --trace-warnings ...` to show where the warning was created)
(node:52635) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag \`--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 2)
^C/home/thate/Documents/utilities/Vesktop/node_modules/electron/dist/electron exited with signal SIGINT

```

</details>

If this can't be changed, then it's completely okay (it's kinda my fault for using a 3-year-old distro too), but I just wanted to let you know in case this is an issue.

RinLovesYou commented:
> It works with the latest commit. The only thing I'd want on my wish list is a way to have mouse bindings, even if I have to do it manually since Discord's in-client keybinding manager doesn't pick it up because of Electron.

A workaround in KDE is the following:

System Settings -> Mouse & Touchpad -> Mouse -> Configure Extra Buttons...

and set it to any keyboard shortcut you don't ever use.

Then you can bind this keyboard combination to discord in the Keyboard Shortcuts settings. This is what i use to do push to talk with one of my extra mouse buttons

tuxinal commented:
> I'm going to be honest. Even after looking at the source code, I could not figure out how I would be able to do that through the CLI.

you have to bind your mouse key's up/down events to `(vesktop) -- --keybind (id) (keyup/keydown)`
note the two sets of dashes
i probably will change the cli at some point so be wary of that i guess

tuxinal commented:
> The global keybinds aren't exposed until you open the keybind settings window in Vencord.

that's intentional. i only tested this on kde which actually opens the keybinds settings page when the register function is called. so i figured it would be better in terms of UX to do it like that.

i'm curious tho do you have to open the keybinds page every time on hyprland for it to work? or is it just the first time? on kde it saves them after the first time.

luna-duclos commented:
A potential bug is that the bindings are registered even when not in a call as long as vesktop is open.
This interacts poorly with, for example, mumble, which will no longer have those keys working for its own binds as long as vesktop is open, then works again once vesktop is closed.
Is it possible to only trigger/register (not sure on the exact mechanism here) them while in a call, or potentially allow them to bubble up ?

allmeta commented:
> > The global keybinds aren't exposed until you open the keybind settings window in Vencord.
> 
> that's intentional. i only tested this on kde which actually opens the keybinds settings page when the register function is called. so i figured it would be better in terms of UX to do it like that.
> 
> i'm curious tho do you have to open the keybinds page every time on hyprland for it to work? or is it just the first time? on kde it saves them after the first time.

Yes, I have to open it on every launch for it to work for that Vesktop session.

CeruleanDerpo commented:
Global shortcuts have just been merged to GNOME
https://gitlab.gnome.org/GNOME/xdg-desktop-portal-gnome/-/issues/47

tuxinal committed:
update venbind

tuxinal committed:
Merge branch 'main' into main

tuxinal commented:
@allmeta you can try with it on hyprland with this latest commit. it worked fine as far as i could tell

it would be nice if anyone could test this on latest gnome too. i think we'd basically be ready to merge after that

A1029384756 commented:
Testing with Gnome on Fedora 42, I did not get the popup to register the shortcut handler. I did test with GPU Screen Recorder and *did* get the popup and Door Knocker reports the portal as functional.

A1029384756 commented:
@Covkie ah, mb. tested with `uwsm` and now i get the dialog on application launch if the key doesn't already exist in dconf

tuxinal committed:
Use xdp registry.
Make keybind IDs string based.
one notable thing is that the default keybind process on discord (as partially used on x11 and windows) does actually have its own ids which are numbers. and so the id strings on x11 and windows are actually _those_ ids as strings
this is opposite to how stuff is on wayland/xdp where the ids are the actual name for the keybind (e.g. TOGGLE_MUTE, PUSH_TO_TALK, ...)

tuxinal committed:
Update the cli
the new cli now looks like `(vesktop) --keybind=(id)` and `(vesktop) --keybind-[up/down]=(id)`
also made the keybinds that get added to xdp a whitelist as opposed to a blacklist of keybinds that shouldn't get added

tuxinal commented:
heads up in the latest commit the cli is now `(vesktop) --keybind=(id)` or `(vesktop) --keybind-[up/down]=(id)`.
the ids are also different (they're now strings). check the keybinds page in settings for all the new ids

tuxinal committed:
bump venbind

StayBlue commented:
> > I'm going to be honest. Even after looking at the source code, I could not figure out how I would be able to do that through the CLI.
> 
> you have to bind your mouse key's up/down events to `(vesktop) -- --keybind (id) ("keyup"/"keydown")` note the two sets of dashes i probably will change the cli at some point so be wary of that i guess

I know on Linux you could probably do this, but on Windows, I'm not sure if you can. Please do it through the CLI on Windows.

tuxinal committed:
Disable Registry usage for now

klhrt commented:
How do the new IDs work? It's still showing up as simply 'ID: 5' same as before, and I can't get any keybinds to trigger anymore.

tuxinal commented:
i forgot to mention the ids are only new on wayland (perhaps because i mainly expected wayland users to use clis). they should be the same as the old ones on x11 (and windows).
does venbind have any errors for you? what is your window manager/desktop environment?

klhrt commented:
I just realized venbind is not building, I wasn't paying attention at all while updating apparently. I'm getting an error stemming from the 'wayland-client' system library. I guess I need to change a flag to prevent venbind from compiling wayland components but I'm not sure how to do so.

WM is a very-barebones i3, no desktop environment or display manager.

tuxinal committed:
make the id patch less ugly, add a description for it
change some texts

tuxinal committed:
Update venbind
loooooootss of changes in the patch related to how keybinds are given to venbind
the main change in venbind is that the api for registring keybinds on xdp and everything else is the same now

tuxinal committed:
Merge remote-tracking branch 'upstream/main'

tuxinal committed:
fix keybinds page not opening on wayland/xdp
i forgor to name things correctly

tuxinal committed:
lint

Razboy20 commented:
@tuxinal Looks like Cmd on MacOS shows as "UNKUNDEFINED":

<img width="704" alt="image" src="https://github.com/user-attachments/assets/de9f9e2a-c324-4e98-bbf2-f88a58dc13d0" />

Additionally, custom keybinds don't seem to be working at all, whether they be unkundefined or have normal modifiers (control, alt, etc.).

Thoughts?

SIGSTACKFAULT commented:
not sure how i got it into this state.

I ran Vesktop through the terminal this morning and global PTT worked just fine. didn't even log out, but when i tried it just now, it didn't work. Deleted the shortcut from KDE System Settings but nothing changed. There's now no trace of Ctrl+\` in my `~/.config`, but it still prints this error when i press Ctrl+\`

```
 git describe --always
12561e3

 npm run start
npm warn Unknown project config "node-linker". This will stop working in the next major version of npm.
npm warn Unknown project config "package-manager-strict". This will stop working in the next major version of npm.

> vesktop@1.5.5 start
> pnpm build && electron .


> vesktop@1.5.5 build /home/tether/Clones/Vesktop
> tsx scripts/build/build.mts

APPIMAGE env is not defined, current application is not an AppImage
Vesktop v1.5.5
checkForUpdatesAndNotify called, downloadPromise is null
[arRPC > ipc] listening at /run/user/1000/discord-ipc-0
[arRPC > websocket] listening on 6463
[arRPC > process] started
(node:4998) UnhandledPromiseRejectionWarning: Error: Cannot read properties of undefined (reading 'keyEvents')
(Use `electron --trace-warnings ...` to show where the warning was created)
(node:4998) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)
(node:4998) UnhandledPromiseRejectionWarning: Error: Cannot read properties of undefined (reading 'keyEvents')
(node:4998) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 2)
(node:4998) UnhandledPromiseRejectionWarning: Error: Cannot read properties of undefined (reading 'keyEvents')
(node:4998) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 3)
(node:4998) UnhandledPromiseRejectionWarning: Error: Cannot read properties of undefined (reading 'keyEvents')
(node:4998) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 4)
(node:4998) UnhandledPromiseRejectionWarning: Error: Cannot read properties of undefined (reading 'keyEvents')
(node:4998) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 5)
(node:4998) UnhandledPromiseRejectionWarning: Error: Cannot read properties of undefined (reading 'keyEvents')
(node:4998) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 6)
^C/home/tether/Clones/Vesktop/node_modules/electron/dist/electron exited with signal SIGINT
```

tuxinal committed:
Change the keybind settings UI on windows and x11
stop using discord's built in keybind store. keybinds are now stored in vesktop's local storage
record shortcuts through venbind instead of electron

tuxinal commented:
i've changed the way keybinds are recorded in the settings page. now it records things through venbind instead of electron (_which does mean it's now even more broken on macos since there is no mac support in venbind_ but if anyone were to make a pr for venbind for mac support it should be slightly easier to fix stuff (@Razboy20))

tuxinal committed:
update venbind. add linux and windows arm64 support
change some styles for keybind settings ui

tuxinal committed:
Merge remote-tracking branch 'upstream/main'

SIGSTACKFAULT commented:
Been using this fork for the past few weeks. works fine. <3

Sycariae commented:
Recent commits have made it dysfunctional, where you can't even save keybinds. Last known working commit: [12561e3](https://github.com/tuxinal/Vesktop/tree/12561e3441ca05b73cd114eacc9c6b84b711403d)

tuxinal committed:
Fix keybinds not saving

tuxinal commented:
I don't even know how I missed that. fixed!

Sycariae commented:
Newest commit fixed the issue I was having, thanks!

I just wanted to ask quick whether there's going to be support for more than 2 keys for a keybind in the future tho?
Also I can't lie I preferred the old UI from [12561e3].(https://github.com/tuxinal/Vesktop/tree/12561e3441ca05b73cd114eacc9c6b84b711403d)

tuxinal commented:
> support for more than 2 keys for a keybind

I was just working on that. if you update the venbind version in package.json to 0.1.6 you _might_ be able to get it working already.
though i'm having some weird issues with it not being able to parse keybinds _only when ran alongside vesktop._ it works fine in every other scenario i genuinely can't tell why this is happening
i'd appreciate if you (or anyone else here) would try this and see if the same thing happens (this is happening on plasma x11 for me)

> I preferred the old UI

yeah that's the original discord ui. i decided to rewrite the whole thing because hacking stuff onto discord's own code sounded like too much of a hassle especially regarding recording keybinds

Sycariae commented:
> i'd appreciate if you (or anyone else here) would try this and see if the same thing happens (this is happening on plasma x11 for me)

Just tried it and not working here either unfortunately (I'm on Windows 11)

tuxinal committed:
update venbind
actually load all the arm preloads

tuxinal committed:
lint

tuxinal committed:
Merge remote-tracking branch 'upstream/main'

tuxinal committed:
update venbind
idk why but pnpm is _really_ insistent on updating es-abstract too for some reason

idarkkz commented:
Hi everyone, I managed to add two global shortcuts: **Mute** and **Deafen**.

Using `globalShortcut`, I injected a JavaScript snippet into the Vesktop window that clicks the specific voice control buttons based on their `[aria-label]` attributes:

- `button[aria-label="Mute"], button[aria-label="Unmute"]`
- `button[aria-label="Deafen"], button[aria-label="Undeafen"]`

I’d like to mention I’m a complete beginner and initially tried building this as a plugin to manage the keybinds, but ran into IPC bridge limitations. I’ll definitely keep working on that.
I'm currently on Linux Fedora KDE Plasma (Wayland).

Here’s what I edited under src/main/index.ts:

```
/*
 * Vencord, a Discord client mod
 * Copyright (c) 2025 Vendicated and contributors
 * SPDX-License-Identifier: GPL-3.0-or-later
 */

import { app, BrowserWindow, globalShortcut, ipcMain, net, protocol } from "electron";
import { join } from "path";
import { pathToFileURL } from "url";

import { initCsp } from "./csp";
import { ensureSafePath } from "./ipcMain";
import { RendererSettings } from "./settings";
import { IS_VANILLA, THEMES_DIR } from "./utils/constants";
import { installExt } from "./utils/extensions";

let muteKey = "Insert";
let deafenKey = "End";

// Register global shortcuts with current keys
function bindGlobalKeybinds() {
    globalShortcut.unregisterAll();

    const win = BrowserWindow.getAllWindows()[0];
    if (!win) {
        console.warn("[GlobalKeybind] No window available for keybinds");
        return;
    }

    globalShortcut.register(muteKey, () => {
        console.log(`[GlobalKeybind] '${muteKey}' pressed. Triggering Mute`);
        win.webContents.executeJavaScript(`
            (() => {
                const btn = document.querySelector('button[aria-label="Mute"], button[aria-label="Unmute"]');
                if (btn) {
                    btn.click();
                    console.log("[GlobalKeybind] ✅ Mute button clicked");
                } else {
                    console.warn("[GlobalKeybind] ❌ Mute/Unmute button not found");
                }
            })();
        `).catch(err => {
            console.error("[GlobalKeybind] Mute JS injection failed:", err);
        });
    });

    globalShortcut.register(deafenKey, () => {
        console.log(`[GlobalKeybind] '${deafenKey}' pressed. Triggering Deafen`);
        win.webContents.executeJavaScript(`
            (() => {
                const btn = document.querySelector('button[aria-label="Deafen"], button[aria-label="Undeafen"]');
                if (btn) {
                    btn.click();
                    console.log("[GlobalKeybind] ✅ Deafen button clicked");
                } else {
                    console.warn("[GlobalKeybind] ❌ Deafen/Undeafen button not found");
                }
            })();
        `).catch(err => {
            console.error("[GlobalKeybind] Deafen JS injection failed:", err);
        });
    });

    console.log(`[GlobalKeybind] Registered: Mute=${muteKey}, Deafen=${deafenKey}`);
}

// Allow plugin to update keys via IPC
ipcMain.handle("globalKeybinds:update", (_event, keys: { mute: string; deafen: string; }) => {
    console.log("[GlobalKeybind] Received new keybinds from plugin:", keys);
    muteKey = keys.mute;
    deafenKey = keys.deafen;
    bindGlobalKeybinds();
});

if (IS_VESKTOP || !IS_VANILLA) {
    app.whenReady().then(() => {
        protocol.handle("vencord", ({ url: unsafeUrl }) => {
            let url = decodeURI(unsafeUrl).slice("vencord://".length).replace(/\?v=\d+$/, "");
            if (url.endsWith("/")) url = url.slice(0, -1);

            if (url.startsWith("/themes/")) {
                const theme = url.slice("/themes/".length);
                const safeUrl = ensureSafePath(THEMES_DIR, theme);
                if (!safeUrl) {
                    return new Response(null, { status: 404 });
                }
                return net.fetch(pathToFileURL(safeUrl).toString());
            }

            switch (url) {
                case "renderer.js.map":
                case "vencordDesktopRenderer.js.map":
                case "preload.js.map":
                case "vencordDesktopPreload.js.map":
                case "patcher.js.map":
                case "vencordDesktopMain.js.map":
                    return net.fetch(pathToFileURL(join(__dirname, url)).toString());
                default:
                    return new Response(null, { status: 404 });
            }
        });

        try {
            if (RendererSettings.store.enableReactDevtools)
                installExt("fmkadmapgofadopljbjfkapdkoienihi")
                    .then(() => console.info("[Vencord] Installed React Developer Tools"))
                    .catch(err => console.error("[Vencord] Failed to install React Developer Tools", err));
        } catch { }

        bindGlobalKeybinds();
        initCsp();
    });
}

if (IS_DISCORD_DESKTOP) {
    require("./patcher");
}

```

Mylloon commented:
is it possible to update the branch?

tuxinal committed:
Merge remote-tracking branch 'upstream/main'

tuxinal commented:
updated @Mylloon

A1029384756 commented:
@SIGSTACKFAULT that is because cosmic does not yet support the global shortcuts portal

marek2001 commented:
After last release only global binding seem to by last high priority feature missing from discord.

PavelDobCZ23 commented:
> I'm asking myself the same thing. I've been using this fork for a while now with no issues (besides causing a systemd-corecump on initial startup). 
> 
@noel-personal 
The coredump is caused by a missing dependency for an Electoron feature that Discord uses. You need to install speech-dispatcher.

SIGSTACKFAULT commented:
> Clearly I'm missing something. How do you register hotkeys in Windows 11 settings?

you don't, that concept is a linux thing. try just binding to something in discord and see if that works.

tuxinal commented:
Hello. this pull request will not currently work with discord as some of its patches are outdated and need to be fixed. 
the lack of contribution is due to vesktop maintainers seemingly having no interest in actually merging this pr. i personally don't want to bother keeping things up to date if no one actually wants to merge anything or review any of the code (heck some even go as far as pretending this doesn't exist at all)

Lonsfor commented:
hearing that they arent interested in merging is actually insane to me

alterNERDtive commented:
Well, personally I don’t think it’s that insane. I have a single use case for this, muting my mic in Discord but not anywhere else. Happens … IDK, twice a year?

nixRidge commented:
I can't imagine it's a widely used feature, no.
I don't know what the reason for the project maintainers not merging this is, all I know/assume is they're 1) stretched pretty thin 2) don't have a lot of time.

For what it's worth, I can still apply and use this patch perfectly fine, it still works fine in KDE Plasma. The shortcuts are registered via the DE, and I can use them as normal.
It'd be convenient if it was merged for general users to use, but I wouldn't go as far as to say it's _insane_ that it hasn't been merged yet.

Has anyone tried to reach out to the developers to get it merged perhaps?

ZeroKun265 commented:
I play a lot of games that have built in voice chat, so I often have to mute myself there and then unmute on discord when I need to talk to everyone one 

Also, I have 4 housemates, so I often need to mute to talk to them if they ask stuff

Toggle mute and deafen for me are an absolutely needed feature. I had been using hyprland for a while and therefore has a way to send shortcuts to certain apps, now that I'm back on plasma I don't, this is a needed feature for me

ZeroKun265 commented:
I play a lot of games that have built in voice chat, so I often have to mute myself there and then unmute on discord when I need to talk to everyone one 

Also, I have 4 housemates, so I often need to mute to talk to them if they ask stuff

Toggle mute and deafen for me are an absolutely needed feature. I had been using hyprland for a while and therefore has a way to send shortcuts to certain apps, now that I'm back on plasma I don't, this is a needed feature for me

jacklollz2 commented:
I'd like to use push-to-mute for in game VC

Mylloon commented:
Should venbind be merged into libvesktop for this PR to be mergeable?

Damglador commented:
Something tells me https://github.com/Vencord/Vesktop/pull/1238 will be merged instead. Which is lame as Electron's support for global shortcuts on Wayland is an afterthought and until Electron 41 or Electron 42 they weren't even labeled properly https://github.com/electron/electron/issues/49873.

yaomtc commented:
@Damglador and that wouldn't resolve the issue of PTT-only voice channels! The only solution there is just to not use Vesktop for them, I guess.

Ketoro commented:
whats stopping this from being merged?

Malix-Labs commented:
https://github.com/Vencord/Vesktop/pull/326#issuecomment-4209769108 from @Ketoro :
> whats stopping this from being merged?

#1238 being worked on

https://github.com/Vencord/Vesktop/pull/326#issuecomment-4196830325 from @Damglador :
> Electron's support for global shortcuts on Wayland is an afterthought

It seems their API has rather stabilized.

It is still better to rely on chromium's native API when possible

Vendicated commented:
Indeed, Chromium's shortcut support on Wayland has really improved now, so it is a viable option and much simpler alternative to this PR. So I have decided to move forward with that instead and plan to finalise and add it in the next version of Vesktop

I really appreciate the work that has been done here but in hindsight it was never really realistic for this to be merged. I am (for the most part) the sole maintainer of Vesktop. This pr is a pretty big change that introduces a new native dependency written in Rust, which I'm not well versed with. Hence, this would add a big burden on me as I would have to actually maintain this library going forward. It's out of scope for me, especially considering there's a viable Electron-native alternative now

So unfortunately yeah this PR won't be accepted. But thank you a lot again Tuxinal, and I'm sorry for not communicating this more clearly earlier!
